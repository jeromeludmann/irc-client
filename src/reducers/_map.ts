import { Action } from "redux";

export const mapReducer = <S, E = void>(map: {
  [action: string]: (state: S, action: Action, extraStates?: E) => S;
}) => (state: S, action: Action, extraStates?: E): S =>
  map.hasOwnProperty(action.type)
    ? map[action.type](state, action, extraStates)
    : state;
