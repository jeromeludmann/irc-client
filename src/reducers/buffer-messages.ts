import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RAW_MESSAGES_RECEIVED,
  RawMessagesReceivedAction,
} from "@app/actions/network";
import {
  JoinAction,
  ServerNoticeAction,
  ServerPingAction,
  ChannelPrivmsgAction,
  UserPrivmsgAction,
  JOIN,
  SERVER_PING,
  CHANNEL_PRIVMSG,
  USER_PRIVMSG,
  SERVER_NOTICE,
} from "@app/actions/message-in";

export type MessageListState = string[];

export type MessageListAction =
  | JoinAction
  | ServerNoticeAction
  | ServerPingAction
  | ChannelPrivmsgAction
  | UserPrivmsgAction
  | ConnectionFailedAction
  | RawMessagesReceivedAction;

export const messagesInitialState: MessageListState = [];

export default (
  messages = messagesInitialState,
  action: MessageListAction,
): MessageListState => {
  switch (action.type) {
    case JOIN: {
      const { user, channel } = action.payload;
      const msg = `${user.nick} has joined ${channel}`;
      return [...messages, msg];
    }
    case SERVER_PING: {
      return [...messages, "Ping? (Pong!)"];
    }
    case CHANNEL_PRIVMSG:
    case USER_PRIVMSG: {
      const { user, text } = action.payload;
      const msg = `${user.nick}: ${text}`;
      return [...messages, msg];
    }
    case SERVER_NOTICE: {
      return [...messages, action.payload.text];
    }
    case CONNECTION_FAILED: {
      return [...messages, action.payload.message];
    }
    case RAW_MESSAGES_RECEIVED: {
      return [...messages, ...action.payload.messages];
    }
    default:
      return messages;
  }
};
