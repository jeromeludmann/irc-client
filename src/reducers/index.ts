import {
  RouteState,
  routeInitialState,
  reduceRoute,
} from "@app/reducers/route";
import {
  ServersState,
  serversInitialState,
  reduceServers,
} from "@app/reducers/servers";
import { RoutedAction } from "@app/Route";
import { Reducer } from "redux";

export interface AppState {
  readonly servers: ServersState;
  readonly route: RouteState;
}

export const rootInitialState = {
  servers: serversInitialState,
  route: routeInitialState,
};

export const reduce: Reducer<AppState, RoutedAction> = (
  state = rootInitialState,
  action,
) => ({
  servers: reduceServers(state.servers, action, { route: state.route }),
  route: reduceRoute(state.route, action, { servers: state.servers }),
});
