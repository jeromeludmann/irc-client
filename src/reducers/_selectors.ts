import { createSelector } from "reselect";
import { RootState } from "@app/reducers";
import { ServersState } from "@app/reducers/servers";
import { RouteState } from "@app/reducers/route";

export const selectServers = ({ servers }: RootState): ServersState => servers;

export const selectRoute = ({ route }: RootState): RouteState => route;

export const selectServer = createSelector(
  selectServers,
  selectRoute,
  (servers, { serverKey }) => servers[serverKey],
);
