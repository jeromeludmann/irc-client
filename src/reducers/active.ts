import { Route, STATUS_WINDOW } from "@app/Route";
import { SwitchWindowAction, SWITCH_WINDOW } from "@app/actions/ui-window";
import { JoinAction, JOIN } from "@app/actions/message-in";
import { UserState } from "@app/reducers/user";

export type ActiveWindowState = Route;

export type ActiveWindowAction = JoinAction | SwitchWindowAction;

export const activeInitialState: ActiveWindowState = {
  server: "serverKey", // TODO multi servers
  window: STATUS_WINDOW,
};

type ExtraParams = {
  user: UserState;
};

export default (
  active = activeInitialState,
  action: ActiveWindowAction,
  { user }: ExtraParams,
): ActiveWindowState => {
  switch (action.type) {
    case JOIN:
      return action.payload.user.nick === user.nick ? action.route : active;

    case SWITCH_WINDOW:
      return action.route;

    default:
      return active;
  }
};
