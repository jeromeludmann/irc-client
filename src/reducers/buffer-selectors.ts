import { createSelector } from "reselect";
import { selectWindow } from "@app/reducers/server-selectors";

export const selectMessages = createSelector(
  selectWindow,
  window => window.messages,
);

export const selectInput = createSelector(selectWindow, window => window.input);
