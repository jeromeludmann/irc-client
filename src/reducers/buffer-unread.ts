import { WindowState } from "@app/reducers/window";
import { Route } from "@app/Route";
import { SWITCH_WINDOW, SwitchWindowAction } from "@app/actions/ui-window";
import {
  RawMessagesReceivedAction,
  RAW_MESSAGES_RECEIVED,
} from "@app/actions/network";
import { UserState } from "@app/reducers/user";
import {
  IncomingJoinAction,
  IncomingChannelPrivmsgAction,
  IncomingChannelNoticeAction,
  JOIN,
  CHANNEL_NOTICE,
  CHANNEL_PRIVMSG,
} from "@app/actions/message-in";

export type UnreadState = boolean;

export type UnreadAction =
  | IncomingJoinAction
  | SwitchWindowAction
  | RawMessagesReceivedAction
  | IncomingChannelPrivmsgAction
  | IncomingChannelNoticeAction;

interface ExtraParams {
  readonly user: UserState;
  readonly active: WindowState;
  readonly route: Route;
}

export const unreadInitialState: UnreadState = false;

export default (
  unread = unreadInitialState,
  action: UnreadAction,
  { user, active, route }: ExtraParams,
): UnreadState => {
  switch (action.type) {
    case SWITCH_WINDOW:
      return false;

    case JOIN:
      return action.payload.user.nick !== user.nick;

    case RAW_MESSAGES_RECEIVED:
    case CHANNEL_NOTICE:
    case CHANNEL_PRIVMSG:
      return (
        active.serverKey !== route.serverKey ||
        active.bufferKey !== route.bufferKey
      );

    default:
      return unread;
  }
};
