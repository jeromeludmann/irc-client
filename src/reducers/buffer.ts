import { Action } from "redux";
import messages, {
  MessageListState,
  MessageListAction,
  messagesInitialState,
} from "@app/reducers/buffer-messages";
import input, {
  InputState,
  InputAction,
  inputInitialState,
} from "@app/reducers/input";
import unread, {
  UnreadState,
  UnreadAction,
  unreadInitialState,
} from "@app/reducers/buffer-unread";
import { WindowState } from "@app/reducers/window";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";

export interface BufferState {
  readonly messages: MessageListState;
  readonly input: InputState;
  readonly unread: UnreadState;
}

export type BufferAction = Action;

interface ExtraParams {
  user: UserState;
  active: WindowState;
  route: Route;
}

export const bufferInitialState: BufferState = {
  messages: messagesInitialState,
  input: inputInitialState,
  unread: unreadInitialState,
};

export default (
  buffer = bufferInitialState,
  action: BufferAction,
  params: ExtraParams,
): BufferState => ({
  messages: messages(buffer.messages, action as MessageListAction),
  input: input(buffer.input, action as InputAction),
  unread: unread(buffer.unread, action as UnreadAction, params),
});
