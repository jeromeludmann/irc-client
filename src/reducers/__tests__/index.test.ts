import { reduce, rootInitialState } from "@app/reducers";

describe("root reducer", () => {
  test("initializing without state", () => {
    expect(
      reduce(undefined, {
        type: "",
        route: { serverKey: "server1", channelKey: "" },
      }),
    ).toMatchSnapshot();
  });

  test("initializing with state", () => {
    expect(
      reduce(rootInitialState, {
        type: "",
        route: { serverKey: "server1", channelKey: "" },
      }),
    ).toMatchSnapshot();
  });
});
