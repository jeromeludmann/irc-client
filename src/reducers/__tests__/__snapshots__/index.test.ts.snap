// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce root state should handle CLOSE_WINDOW on channel 1`] = `
Object {
  "route": Object {
    "bufferKey": "@status",
    "serverKey": "",
  },
  "servers": Object {},
}
`;

exports[`reduce root state should handle CLOSE_WINDOW on private 1`] = `
Object {
  "route": Object {
    "bufferKey": "@status",
    "serverKey": "",
  },
  "servers": Object {},
}
`;

exports[`reduce root state should handle CLOSE_WINDOW on status with at least two servers 1`] = `
Object {
  "route": Object {
    "bufferKey": "@status",
    "serverKey": "serverKey1",
  },
  "servers": Object {
    "serverKey1": Object {
      "buffers": Object {
        "@raw": Object {
          "activity": false,
          "input": Object {
            "dirtyValue": "",
            "history": Object {
              "index": 0,
              "values": Array [],
            },
            "value": "",
          },
          "messages": Array [],
        },
        "@status": Object {
          "activity": false,
          "input": Object {
            "dirtyValue": "",
            "history": Object {
              "index": 0,
              "values": Array [],
            },
            "value": "",
          },
          "messages": Array [],
        },
      },
      "lag": 0,
      "modes": Object {
        "available": Object {
          "channel": Array [],
          "user": Array [],
        },
        "user": Array [],
      },
      "name": "<unknown>",
      "user": Object {
        "nick": "default_nick",
        "real": "default_name",
        "user": "default_user",
      },
    },
    "serverKey2": Object {
      "buffers": Object {
        "@raw": Object {
          "activity": false,
          "input": Object {
            "dirtyValue": "",
            "history": Object {
              "index": 0,
              "values": Array [],
            },
            "value": "",
          },
          "messages": Array [],
        },
        "@status": Object {
          "activity": false,
          "input": Object {
            "dirtyValue": "",
            "history": Object {
              "index": 0,
              "values": Array [],
            },
            "value": "",
          },
          "messages": Array [],
        },
      },
      "lag": 0,
      "modes": Object {
        "available": Object {
          "channel": Array [],
          "user": Array [],
        },
        "user": Array [],
      },
      "name": "<unknown>",
      "user": Object {
        "nick": "default_nick",
        "real": "default_name",
        "user": "default_user",
      },
    },
  },
}
`;

exports[`reduce root state should handle CONNECT_TO_SERVER action 1`] = `
Object {
  "route": Object {
    "bufferKey": "@status",
    "serverKey": "",
  },
  "servers": Object {
    "serverKey": Object {
      "buffers": Object {
        "@raw": Object {
          "activity": false,
          "input": Object {
            "dirtyValue": "",
            "history": Object {
              "index": 0,
              "values": Array [],
            },
            "value": "",
          },
          "messages": Array [],
        },
        "@status": Object {
          "activity": false,
          "input": Object {
            "dirtyValue": "",
            "history": Object {
              "index": 0,
              "values": Array [],
            },
            "value": "",
          },
          "messages": Array [
            "Connecting to serverâ€¦",
          ],
        },
      },
      "lag": 0,
      "modes": Object {
        "available": Object {
          "channel": Array [],
          "user": Array [],
        },
        "user": Array [],
      },
      "name": "<unknown>",
      "user": Object {
        "nick": "default_nick",
        "real": "default_name",
        "user": "default_user",
      },
    },
  },
}
`;

exports[`reduce root state should not handle anything 1`] = `
Object {
  "route": Object {
    "bufferKey": "@status",
    "serverKey": "",
  },
  "servers": Object {},
}
`;
